{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "DataPreprocessingCodons.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "BiENFpNI7YbL"
      },
      "source": [
        "!pip install bio\n",
        "from Bio import SeqIO\n",
        "from google.colab import drive\n",
        "import numpy as np\n",
        "# drive.mount('/content/drive')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Izwh3Uy972EP"
      },
      "source": [
        "# Read fasta into a list\n",
        "fasta_file = \"/content/drive/My Drive/sample_reads.fasta\"\n",
        "fasta_name = []\n",
        "sequence_list = []\n",
        "#\n",
        "for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n",
        "    sequence_list.append(str(seq_record.seq))\n",
        "    fasta_name.append(seq_record.id)\n",
        "\n",
        "# print(np.array(sequence_list))"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hkgbNOu2_BfE",
        "outputId": "5090d889-7c9c-4a8b-a8c9-21840194e1d9"
      },
      "source": [
        "# Transform data from N batches of L-long fragments (N x L)\n",
        "# into N x 3 x (L // 3) x 21 matrix\n",
        "# (3 different start positions, 21 CAT combinations in a ones hot matrix)\n",
        "\n",
        "# Enumerate amino acids for ones-hot index\n",
        "ALANINE = 0\n",
        "ASPARAGINE = 1\n",
        "ASPARTIC_ACID = 2\n",
        "ARGININE = 3\n",
        "CYSTEINE = 4\n",
        "GLUTAMIC_ACID = 5\n",
        "GLUTAMINE = 6\n",
        "GLYCINE = 7\n",
        "HISTIDINE = 8\n",
        "ISOLEUCINE = 9\n",
        "LEUCINE = 10\n",
        "LYSINE = 11\n",
        "METHIONINE = 12\n",
        "PHENYLALANINE = 13\n",
        "PROLINE = 14\n",
        "SERINE = 15\n",
        "THREONINE = 16\n",
        "TRYPTOPHAN = 17\n",
        "TYROSINE = 18\n",
        "VALINE = 19\n",
        "STOP = 20  # should always be last\n",
        "\n",
        "# Amino acids sharing codon with the stop codons\n",
        "# SELENOCYSTEINE = 20\n",
        "# PYRROLYSINE = 21\n",
        "\n",
        "# Dictionary mapping codon to amino acids\n",
        "dna_codon_mapping = {\n",
        "    \"GCT\": ALANINE,\n",
        "    \"GCC\": ALANINE,\n",
        "    \"GCA\": ALANINE,\n",
        "    \"GCG\": ALANINE,\n",
        "    \"AAT\": ASPARAGINE,\n",
        "    \"AAC\": ASPARAGINE,\n",
        "    \"GAT\": ASPARTIC_ACID,\n",
        "    \"GAC\": ASPARTIC_ACID,\n",
        "    \"CGT\": ARGININE,\n",
        "    \"CGC\": ARGININE,\n",
        "    \"CGA\": ARGININE,\n",
        "    \"CGG\": ARGININE,\n",
        "    \"AGA\": ARGININE,\n",
        "    \"AGG\": ARGININE,\n",
        "    \"TGT\": CYSTEINE,\n",
        "    \"TGC\": CYSTEINE,\n",
        "    \"GAA\": GLUTAMIC_ACID,\n",
        "    \"GAG\": GLUTAMIC_ACID,\n",
        "    \"CAA\": GLUTAMINE,\n",
        "    \"CAG\": GLUTAMINE,\n",
        "    \"GGT\": GLYCINE,\n",
        "    \"GGC\": GLYCINE,\n",
        "    \"GGA\": GLYCINE,\n",
        "    \"GGG\": GLYCINE,\n",
        "    \"CAT\": HISTIDINE,\n",
        "    \"CAC\": HISTIDINE,\n",
        "    \"ATT\": ISOLEUCINE,\n",
        "    \"ATC\": ISOLEUCINE,\n",
        "    \"ATA\": ISOLEUCINE,\n",
        "    \"TTA\": LEUCINE,\n",
        "    \"TTG\": LEUCINE,  # can also be a start codon\n",
        "    \"CTT\": LEUCINE,\n",
        "    \"CTC\": LEUCINE,\n",
        "    \"CTA\": LEUCINE,\n",
        "    \"CTG\": LEUCINE,\n",
        "    \"AAA\": LYSINE,\n",
        "    \"AAG\": LYSINE,\n",
        "    \"ATG\": METHIONINE,  # can also be a start codon\n",
        "    \"TTT\": PHENYLALANINE,\n",
        "    \"TTC\": PHENYLALANINE,\n",
        "    \"CCT\": PROLINE,\n",
        "    \"CCC\": PROLINE,\n",
        "    \"CCA\": PROLINE,\n",
        "    \"CCG\": PROLINE,\n",
        "    \"TCT\": SERINE,\n",
        "    \"TCC\": SERINE,\n",
        "    \"TCA\": SERINE,\n",
        "    \"TCG\": SERINE,\n",
        "    \"AGT\": SERINE,\n",
        "    \"AGC\": SERINE,\n",
        "    \"ACT\": THREONINE,\n",
        "    \"ACC\": THREONINE,\n",
        "    \"ACA\": THREONINE,\n",
        "    \"ACG\": THREONINE,\n",
        "    \"TGG\": TRYPTOPHAN,\n",
        "    \"TAT\": TYROSINE,\n",
        "    \"TAC\": TYROSINE,\n",
        "    \"GTT\": VALINE,\n",
        "    \"GTC\": VALINE,\n",
        "    \"GTA\": VALINE,\n",
        "    \"GTG\": VALINE,  # can also be a start codon\n",
        "    \n",
        "    # Encoding stop codons together (rather than as independent amino acids)\n",
        "    \"TGA\": STOP,\n",
        "    \"TAG\": STOP,\n",
        "    \"TAA\": STOP,\n",
        "}\n",
        "\n",
        "\n",
        "# Function to iterate over a fragment to build the (3 x (L // 3) x 21 matrix)\n",
        "def encode_fragment(fragment):\n",
        "    result = np.zeros((3, (len(fragment) // 3), STOP + 1))\n",
        "\n",
        "    for i in range(len(fragment) - 2):\n",
        "\n",
        "        # 'N' means the base could not be identified due to DNA sequence quality, ignore\n",
        "        # Maybe should remove the N from the sequence before encoding (scrubbing)\n",
        "        if 'N' in fragment[i:i+3]:\n",
        "          continue\n",
        "        codon = dna_codon_mapping[fragment[i:i+3]]\n",
        "\n",
        "        # i % 3 selects which of the 3 start position fragments to set\n",
        "        # i // 3 slides forward in each fragment after setting for each start position\n",
        "        # codon indexes the ones-hot\n",
        "        result[i % 3][i // 3][codon] = 1\n",
        "    \n",
        "    return result\n",
        "\n",
        "encoded_sequence_array = np.zeros((len(sequence_list), 3, len(sequence_list[0]) // 3, STOP + 1))\n",
        "print(encoded_sequence_array.shape, len(sequence_list), len(sequence_list[0]))\n",
        "\n",
        "for i, frag in enumerate(sequence_list):\n",
        "    encoded_sequence_array[i] = encode_fragment(frag)\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(62500, 3, 83, 21) 62500 250\n"
          ]
        }
      ]
    }
  ]
}